Update power management page (#111)
dortania/OpenCore-Post-Install: universal/pm.md

## Enabling X86PlatformPlugin

So before we can fine tune power management to our liking, we need to first make sure Apple's XCPM core is loaded. Note that this is supported **only on Haswell and newer**, consumer Sandy, Ivy Bridge and AMD CPUs should refer to the following:

* [Sandy and Ivy Bridge Power Management](../universal/pm.md#sandy-and-ivy-bridge-power-management)
* [AMD CPU Power Management](../universal/pm.md#amd-cpu-power-management)

::: details Ivy Bridge and Ivy Bridge-E note

> Apple dropped support for XCPM back in macOS Sierra, so XCPM is only supported between 10.8.5 and 10.11.6. Newer OSes will require the [ssdtPRgen method](../universal/pm.md#sandy-and-ivy-bridge-power-management)

Reviewer: ssdtPRGen will always be required regardless of AICPUPM or XCPM.

> Apple dropped support for XCPM on these models back in macOS Sierra, so XCPM is only supported between 10.8.5 and 10.11.6. You will still need [ssdtPRgen](../universal/pm.md#sandy-and-ivy-bridge-power-management).

Reviewer: Just to be precise, if I am not too pedantic - `-xcpm` was only used on Ivy.

:::

::: tip

To enable XCPM in 10.11 and older, simply add `-xcpm` to your boot-args.

:::

> To start, grab [IORegistryExplorer](https://github.com/khronokernel/IORegistryClone/blob/master/ioreg-302.zip) and look for `AppleACPICPU` (note that if you use search, it won't show the children so clear the search box once you've found the entry):

Reviewer 2: ```suggestion
To start, grab [IORegistryExplorer](https://github.com/khronokernel/IORegistryClone/blob/master/ioreg-302.zip) and look for `AppleACPICPU` (Note that if you use search, IORegistryExplorer won't show the children of services found. Be sure to clear the search box once you've found the entry):
```

XCPM Present           |  Missing XCPM
:-------------------------:|:-------------------------:
![](../images/post-install/pm-md/pm-working.png)  |  ![](../images/post-install/pm-md/pm-not-working.png)

> As you can see from the image on the left, we have X86PlatformPlugin attached meaning Apple's CPU power management drivers are working (note that the name of the CPU does not matter). If you see something similar to the image on the right, then there's likely an issue. Make sure to check the following:

Reviewer 2: ```suggestion
As you can see from the image on the left, we have X86PlatformPlugin attached. This means that Apple's CPU power management drivers are working (Note that the name of the CPU does not matter). If you see something similar to the image on the right, then there's likely an issue. Make sure to check the following:
```

> As you can see from the image on the left, we have X86PlatformPlugin attached meaning Apple's CPU power management drivers are doing their thing (note that the name of the CPU does not matter). If you see something similar to the image on the right, then there's likely an issue. Make sure to check the following:

Reviewer: doing their thingS?

* SSDT-PLUG.**aml** is both present and enabled in your config.plist and EFI/OC/ACPI
  * If you're missing this, head to [Getting Started With ACPI](https://dortania.github.io/Getting-Started-With-ACPI) on how to make this

> * SSDT-PLUG is set to the first thread of your CPU, you can check by selecting the first CPU listed (`CP00` for our example) and make sure you see this in the properties:

Reviewer 2: ```suggestion
* SSDT-PLUG is set to the first thread of your CPU. You can check by selecting the first CPU listed (`CP00` for our example) and make sure you see this in the properties:
```

> * SSDT-PLUG is set to the first thread of your CPU. You can check by selecting the first CPU listed (`CP00` for our example) and make sure you see this in the properties:

Reviewer: Shall we mention that SSDT-PLUG is no longer required on 12.4+? Or is it already included elsewhere?

Author: Will mention it in the ACPI guide too, although best to mention it in both places to not confuse people.

```
plugin-type | Number | 0x1
```

::: details X99 Note

> XCPM does not natively support Haswell-E and Broadwell-E, this means we need to spoof the CPU ID into a model that does supports XCPM:

Reviewer: Uh. I think Haswell-E is supported by `_cpuid_set_info`, so it will not KP. Broadwell-E is not supported at all.

Better to mention this, though I do not really mind, as we do need to do the CPUID spoofing for both generations.

XCPM Present           |  Missing XCPM
:-------------------------:|:-------------------------:
![](../images/post-install/pm-md/pm-working.png)  |  ![](../images/post-install/pm-md/pm-not-working.png)

As you can see from the image on the left, we have X86PlatformPlugin attached. This means that Apple's CPU power management drivers are working (the name of the CPU does not matter). If you see something similar to the image on the right, then there's likely an issue. Make sure to check the following:

* SSDT-PLUG.**aml** is both present and enabled in your config.plist and EFI/OC/ACPI
  * If you're missing this, head to [Getting Started With ACPI](https://dortania.github.io/Getting-Started-With-ACPI) on how to make this
* SSDT-PLUG is set to the first thread of your CPU. You can check by selecting the first CPU listed (`CP00` for our example) and make sure you see this in the properties:

::: tip

> SSDT-PLUG is not required on 12.4 and up.

Reviewer 3: `plugin-type=1` check was dropped on macOS 12.**3**b1+

Reviewer: Right. Thanks. Please file a PR for this.

> Apple dropped support for XCPM on these models back in macOS Sierra, so XCPM is only supported between 10.8.5 and 10.11.6. You will still need [ssdtPRgen](../universal/pm.md#sandy-and-ivy-bridge-power-management).

Reviewer: Hm, is it actually worth mentioning that XCPM can still be enabled on Ivy Bridge on 10.12+ by patching XNU? If this might confuse users, then better not.

Author: probably will

    * Cpuid1Data: `D4060300 00000000 00000000 00000000`
    * Cpuid1Mask: `FFFFFFFF 00000000 00000000 00000000`

## Manually Modifying Frequency vectors

> In most cases, the native CPU power management data shipped with macOS works out of the box. If you're experiencing issues, changing SMBIOS to something more appropriate to your system will provide different frequency vectors and may be better for your usecase. In the cases where manual tuning is required (unsupported CPU or a CPU not used in any Macs), you can use CPUFriend to inject modified frequency vectors, but if you don't know what you're doing, you can severely break power management.

Reviewer: works -> work (data is plural)

> In most cases, the native CPU power management data shipped with macOS works out of the box. If you're experiencing issues, changing SMBIOS to something more appropriate to your system will provide different frequency vectors and may be better for your usecase. In the cases where manual tuning is required (unsupported CPU or a CPU not used in any Macs), you can use CPUFriend to inject modified frequency vectors, but if you don't know what you're doing, you can severely break power management.

Reviewer: > you can use CPUFriend to inject modified frequency vectors

Wrong. CPUFriend handles the whole power management data, instead of FV only.

> In most cases, the native CPU power management data shipped with macOS works out of the box. If you're experiencing issues, changing SMBIOS to something more appropriate to your system will provide different frequency vectors and may be better for your usecase. In the cases where manual tuning is required (unsupported CPU or a CPU not used in any Macs), you can use CPUFriend to inject modified frequency vectors, but if you don't know what you're doing, you can severely break power management.

Reviewer: > a CPU not used in any Macs

There are only very limited numbers of CPUs used by real Macs. In my opinion, just drop the parentheses; skilled users know what they are doing.

    * Cpuid1Data: `D4060300 00000000 00000000 00000000`
    * Cpuid1Mask: `FFFFFFFF 00000000 00000000 00000000`

:::

## Manually Modifying Power Management Data

In most cases, the native CPU power management data shipped with macOS work out of the box. If you're experiencing issues, changing your SMBIOS to something more appropriate to your system will provide different data and may be better for your usecase. In the cases where manual tuning is required, you can use CPUFriend to inject modified power management data, but if you don't know what you're doing, you can severely break power management.

::: warning

In most cases, you do not have to do this. Change your SMBIOS instead.

:::

::: tip

> This is a example on how to change some parts of power management data. For more information, you should check out [CPUFriend's documentation](https://github.com/acidanthera/CPUFriend/blob/master/Instructions.md).

Reviewer: aN example

:::

### Using CPUFriend

To start, we're gonna need a couple things:

> 

Reviewer: For the line below:

`X86PlatformPlugin` is also used on Ivy Bridge, just with AICPUPM. Should be Sandy Bridge instead I think?

Once you're finished, you'll be provided with a CPUFriendDataProvider.kext and ssdt_data.aml. Which you choose is your preference but I recommend the kext variant to avoid any headaches with data injection into Windows and Linux.

* **Note**: Load order does not matter with the CPUFriendDataProvider as it's just a plist-only kext

> * **Note 2**: Wake issues resulting from CPUFriend is likely due to incorrect power management data, every system is unique so you'll need to play around until you get a stable config. Kernel panics will have `Sleep Wake failure in efi`. Reusing power management data from old macOS versions can also cause issues, so recreate your data provider if you update macOS.

Reviewer 2: ```suggestion
* **Note 2**: Wake issues resulting from CPUFriend are likely due to incorrect power management data. Every system is unique so you'll need to play around until you get a stable config. Kernel panics will have `Sleep Wake failure in efi`. Reusing power management data from old macOS versions can also cause issues, so recreate your data provider if you update macOS.
```

Once you're finished, you'll be provided with a CPUFriendDataProvider.kext and ssdt_data.aml. Which you choose is your preference but I recommend the kext variant to avoid any headaches with data injection into Windows and Linux.

* **Note**: Load order does not matter with the CPUFriendDataProvider as it's just a plist-only kext
* **Note 2**: Wake issues resulting from CPUFriend are likely due to incorrect power management data. Every system is unique so you'll need to play around until you get a stable config. Kernel panics will have `Sleep Wake failure in efi`. Reusing power management data from old macOS versions can also cause issues, so recreate your data provider if you update macOS.

> * **Note 3**: If you do choose to use ssdt_data.aml, note that SSDT-PLUG is no longer needed. However the setup for this SSDT is broken on HEDT platforms like X99 and X299, so we highly recommend SSDT-PLUG with CPUFriendDataProvider.kext instead.

Reviewer: Just an idea following Note 2 - maybe we could mention creating `CPUFriendDataProvider_OSVersion.kext` and then set `Min/MaxKernel` such that specific set of power management data is injected for different versions of macOS.

* Ensure CpuPm and Cpu0Ist tables are **NOT** dropped
* [ssdtPRGen](https://github.com/Piker-Alpha/ssdtPRGen.sh)

> Initially with the setup in the Ivy Bridge section, we recommended users drop their CpuPm and Cpu0Ist to avoid any issues with AppleIntelCPUPowerManagement.kext. But dropping these tables has the adverse affect of breaking turbo boost in Windows. So to resolve this, we'll want to keep our OEM's table but we'll want to add a new table to supplement data only for macOS. So once we're done creating our CPU-PM table, we'll re-add our OEM's CPU SSDTs.

Reviewer: If I remember correctly, there is no need to drop the OEM Cpu0Ist and CpuPm.

Reviewer 4: I can confirm this is still needed to be done until you build the SSDT from ssdtPRgen.  If not, a kernel panic will happen from AppleIntelPowerManagement.

Reviewer: Thanks for the confirmation! Then let's leave it as is. Resolved.
